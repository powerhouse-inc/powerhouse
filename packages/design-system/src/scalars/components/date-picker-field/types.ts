export type DateFieldValue = string | Date | undefined;
export type DatePickerView = "years" | "months" | "days";
export type WeekStartDayNumber = 0 | 1 | 2 | 3 | 4 | 5 | 6 | undefined;
