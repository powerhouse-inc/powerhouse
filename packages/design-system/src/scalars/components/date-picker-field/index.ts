export {
  DatePickerField,
  type DatePickerFieldProps,
} from "./date-picker-field.js";
