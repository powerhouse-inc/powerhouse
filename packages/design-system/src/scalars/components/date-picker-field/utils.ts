import { format, isValid, parse } from "date-fns";
import { DateFieldValue } from "./types";

export const ALLOWED_FORMATS = [
  "yyyy-MM-dd",
  "dd/MM/yyyy",
  "MM/dd/yyyy",
  "dd-MMM-yyyy",
  "MMM-dd-yyyy",
];

export const isFormatAllowed = (dateString: string) =>
  Object.values(dateFormatRegexes).some((regex) => regex.test(dateString));

export const isDateFormatAllowed = (
  dateString: string,
  dateFormat?: string,
) => {
  if (!dateFormat) return isFormatAllowed(dateString);
  const regex = dateFormatRegexes[dateFormat as keyof typeof dateFormatRegexes];
  if (!regex) return false;
  return regex.test(dateString);
};

export const dateFormatRegexes = {
  "yyyy-MM-dd": /^\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])$/,
  "dd/MM/yyyy": /^(0[1-9]|[12]\d|3[01])\/(0[1-9]|1[0-2])\/\d{4}$/,
  "MM/dd/yyyy": /^(0[1-9]|1[0-2])\/(0[1-9]|[12]\d|3[01])\/\d{4}$/,
  "dd-MMM-yyyy":
    /^(0[1-9]|[12]\d|3[01])-(?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)-\d{4}$/,
  "MMM-dd-yyyy":
    /^(?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)-(0[1-9]|[12]\d|3[01])-\d{4}$/,
};

const splitDateTime = (
  value: DateFieldValue,
): { date: string; time: string } => {
  const defaultResult = { date: "", time: "" };

  if (!value) {
    return defaultResult;
  }

  if (typeof value === "string") {
    const [date, time] = value.split("T");
    return { date, time };
  }

  if (isValid(value)) {
    const dateString = value.toISOString();
    const [date, time] = dateString.split("T");
    return { date, time };
  }

  return defaultResult;
};

export const getDateFromValue = (value: DateFieldValue): string => {
  const { date } = splitDateTime(value);
  return date;
};

export const parseInputString = (
  inputString: string,
  dateFormat = ALLOWED_FORMATS[0],
): string => {
  if (!dateFormat || !inputString) return inputString;

  for (const [formatStr, regex] of Object.entries(dateFormatRegexes)) {
    if (regex.test(inputString)) {
      const parsedDate = parse(inputString, formatStr, new Date());
      if (isValid(parsedDate)) {
        const newValue = format(parsedDate, dateFormat || formatStr);
        return newValue;
      }
    }
  }
  return inputString;
};

export const formatDateToValue = (inputString: string): string => {
  // Add the time and the timezone to the date to split it later by T
  const stringDate = `${inputString}T00:00-00:00`;
  return stringDate;
};

/**
 * Convert a date string to an ISO date string (YYYY-MM-DD format)
 * @param inputString - The date string to convert
 * @param dateFormat - The date format to use
 * @returns The ISO date string
 * @example
 * // MM/dd/yyyy format
 * convertToISODateFormat('12/25/2023') // returns '2023-12-25'
 *
 * // dd/MM/yyyy format
 * convertToISODateFormat('25/12/2023') // returns '2023-12-25'
 *
 * // dd-MMM-yyyy format
 * convertToISODateFormat('25-Dec-2023') // returns '2023-12-25'
 *
 * // Invalid date
 * convertToISODateFormat('invalid') // returns 'invalid'
 */
export const convertToISODateFormat = (
  inputString: string,
  dateFormat = ALLOWED_FORMATS[0],
): string => {
  if (!dateFormat || !inputString) return inputString;
  // Si la fecha tiene formato MM/dd/yyyy
  if (dateFormatRegexes["MM/dd/yyyy"].test(inputString)) {
    const [month, day, year] = inputString.split("/");
    // Construimos la fecha ISO en el orden correcto: año-mes-día
    return `${year}-${month.padStart(2, "0")}-${day.padStart(2, "0")}`;
  }

  for (const [formatStr, regex] of Object.entries(dateFormatRegexes)) {
    if (regex.test(inputString)) {
      const parsedDate = parse(inputString, formatStr, new Date());
      if (isValid(parsedDate)) {
        // Formatear a ISO sin tiempo
        const isoDate = format(parsedDate, "yyyy-MM-dd");
        return isoDate;
      }
    }
  }
  return inputString;
};

/**
 * Convert a date string to an ISO date string (YYYY-MM-DD format)
 * @param inputString - The date string to convert
 * @returns The ISO date string
 */
export const formatDateToValidCalendarDateFormat = (
  inputString: string,
): string => {
  const [datePart, timePart = "00:00-00:00"] = inputString.split("T");
  const parsedDate = convertToISODateFormat(datePart);

  // Combinar con la parte del tiempo y convertir a ISO
  const isoDate = new Date(`${parsedDate}T${timePart}`);
  return isoDate.toISOString();
};
