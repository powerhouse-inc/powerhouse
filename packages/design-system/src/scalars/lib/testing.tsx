import { render, type RenderResult } from "@testing-library/react";
import { Form } from "../components/index.js";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
