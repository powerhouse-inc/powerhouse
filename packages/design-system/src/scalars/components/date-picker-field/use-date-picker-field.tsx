import { parse, startOfDay } from "date-fns";
import React, { useCallback, useMemo } from "react";
import { DateFieldValue, WeekStartDayNumber } from "./types";
import {
  formatDateToValue,
  formatUTCDateToISOStringWithOutTime,
  getDateFromValue,
  isDateFormatAllowed,
  parseInputString,
} from "./utils";
import { createChangeEvent } from "../time-picker-field/utils";

interface DatePickerFieldProps {
  value?: DateFieldValue;
  defaultValue?: DateFieldValue;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onBlur?: (e: React.FocusEvent<HTMLInputElement>) => void;
  disablePastDates?: boolean;
  disableFutureDates?: boolean;
  dateFormat?: string;
  weekStart?: string;
  autoClose?: boolean;
}

export const useDatePickerField = ({
  value,
  defaultValue,
  onChange,
  onBlur,
  disablePastDates,
  disableFutureDates,
  dateFormat,
  weekStart = "Monday",
  autoClose = false,
}: DatePickerFieldProps) => {
  const [isOpen, setIsOpen] = React.useState(false);
  const [inputDisplay, setInputDisplay] = React.useState<string | undefined>(
    parseInputString(getDateFromValue(value ?? defaultValue ?? ""), dateFormat),
  );

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const inputValue = e.target.value;
    setInputDisplay(inputValue);
    onChange?.(createChangeEvent(inputValue));
  };

  const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
    onBlur?.(e);
  };

  const handleDateSelect = useCallback(
    (date?: Date) => {
      if (!date) return;

      // Get LOCAL date components
      const localYear = date.getFullYear();
      const localMonth = date.getMonth();
      const localDay = date.getDate();
      // Create UTC date representing the same local date
      const utcDate = new Date(Date.UTC(localYear, localMonth, localDay));
      // Take the date without time in ISO format
      const inputValue = formatUTCDateToISOStringWithOutTime(utcDate);
      // Parse the date to the correct format
      const newInputValue = parseInputString(inputValue, dateFormat);
      setInputDisplay(newInputValue);
      const newValue = formatDateToValue(utcDate);

      const newValueOnchage = `${newInputValue}T${newValue.split("T")[1]}`;

      onChange?.(createChangeEvent(newValueOnchage));
    },
    [dateFormat, onChange],
  );

  const today = useMemo(() => startOfDay(new Date()), []);

  const disabledDates = useMemo(
    () =>
      disablePastDates && disableFutureDates
        ? {
            before: today,
            after: today,
          }
        : disablePastDates
          ? {
              before: today,
            }
          : disableFutureDates
            ? {
                after: today,
              }
            : undefined,
    [disablePastDates, disableFutureDates, today],
  );

  const weekStartDay = useMemo(() => {
    const days = [
      "sunday",
      "monday",
      "tuesday",
      "wednesday",
      "thursday",
      "friday",
      "saturday",
    ];
    const dayIndex = days.indexOf(weekStart.toLowerCase());
    return (dayIndex >= 0 ? dayIndex : 1) as WeekStartDayNumber;
  }, [weekStart]);

  // Close the calendar when a date is selected
  const handleDayClick = () => {
    if (autoClose) {
      setIsOpen(false);
    }
  };

  const date = useMemo(() => {
    if (!value) return undefined;
    const dateString = getDateFromValue(value);
    const isValidDate = isDateFormatAllowed(dateString);
    if (!isValidDate) return undefined;

    const dateStringFormatted = parseInputString(dateString, dateFormat);
    const fechaUTC = parse(
      dateStringFormatted,
      dateFormat ?? "yyyy-MM-dd",
      new Date(),
    );

    return fechaUTC;
  }, [value, dateFormat]);

  return {
    date,
    inputValue: inputDisplay,
    handleDateSelect,
    handleInputChange,
    isOpen,
    setIsOpen,
    handleBlur,
    disabledDates,
    weekStartDay,
    handleDayClick,
  };
};
