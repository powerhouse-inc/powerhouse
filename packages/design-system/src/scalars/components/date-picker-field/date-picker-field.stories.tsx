import type { Meta, StoryObj } from "@storybook/react";
import { withForm } from "@/scalars/lib/decorators";
import DatePickerField from "./date-picker-field";
import { getDefaultArgTypes } from "@/scalars/lib/storybook-arg-types";

const meta: Meta<typeof DatePickerField> = {
  title: "Document Engineering/Simple Components/Date Picker Field",
  component: DatePickerField,
  parameters: {
    layout: "centered",
  },
  decorators: [withForm],
  tags: ["autodocs"],
  argTypes: {
    ...getDefaultArgTypes({
      valueControlType: "date",
      valueType: "date",
    }),
  },
  args: {
    name: "date-picker-field",
    errors: [],
    warnings: [],
  },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    name: "date",
    label: "Pick a date",
    value: "2025-01-27",
  },
};
