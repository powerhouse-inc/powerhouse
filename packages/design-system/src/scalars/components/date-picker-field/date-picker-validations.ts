import { format } from "date-fns";
import { type DateFieldValue } from "../../../ui/components/data-entry/date-picker/types.js";

import {
  formatDateToValidCalendarDateFormat,
  getDateFromValue,
} from "../../../ui/components/data-entry/date-picker/utils.js";
import {
  getDateFormat,
  isDateFormatAllowed,
  normalizeMonthFormat,
} from "../../../ui/components/data-entry/date-time-picker/utils.js";
import { type DatePickerFieldProps } from "./date-picker-field.js";

export const validateDatePicker =
  ({
    dateFormat,
    minDate,
    maxDate,
    disablePastDates,
    disableFutureDates,
  }: DatePickerFieldProps) =>
  (value: unknown) => {
    if (value === "" || value === undefined) {
      return true;
    }

    const internalFormat = getDateFormat(dateFormat ?? "");
    const stringDate = normalizeMonthFormat(
      getDateFromValue(value as DateFieldValue),
    );

    const isValid = isDateFormatAllowed(stringDate, internalFormat);

    if (!isValid) {
      return `Invalid date format. Please use a valid format`;
    }
    const isoDate = formatDateToValidCalendarDateFormat(stringDate);
    // Split the date and time parts
    const [datePart] = isoDate.split("T");
    // Create a valid date object with the date part and set the time to 00:00:00
    const validDateStartOfDay = new Date(`${datePart}T00:00:00`);

    // Get the most restrictive date between minDate and disablePastDates
    let effectiveMinDate: Date | null = null;
    if (minDate) {
      const minDateValue = new Date(minDate);
      effectiveMinDate = new Date(minDateValue.setHours(0, 0, 0, 0));
    }
    if (disablePastDates) {
      const today = new Date();
      const todayStartOfDay = new Date(today.setHours(0, 0, 0, 0));
      if (!effectiveMinDate || todayStartOfDay > effectiveMinDate) {
        effectiveMinDate = todayStartOfDay;
      }
    }

    // Get the most restrictive date between maxDate and disableFutureDates
    let effectiveMaxDate: Date | null = null;
    if (maxDate) {
      const maxDateValue = new Date(maxDate);
      effectiveMaxDate = new Date(maxDateValue.setHours(0, 0, 0, 0));
    }
    if (disableFutureDates) {
      const today = new Date();
      const todayStartOfDay = new Date(today.setHours(0, 0, 0, 0));
      if (!effectiveMaxDate || todayStartOfDay < effectiveMaxDate) {
        effectiveMaxDate = todayStartOfDay;
      }
    }

    // Validate against the effective dates
    if (effectiveMinDate && effectiveMaxDate) {
      if (effectiveMinDate > effectiveMaxDate) {
        const formattedMinDate = format(
          effectiveMinDate,
          internalFormat ?? "yyyy-MM-dd",
        );
        const formattedMaxDate = format(
          effectiveMaxDate,
          internalFormat ?? "yyyy-MM-dd",
        );
        return `Invalid date range: ${formattedMinDate.toUpperCase()} is after ${formattedMaxDate.toUpperCase()}`;
      }

      if (
        validDateStartOfDay < effectiveMinDate ||
        validDateStartOfDay > effectiveMaxDate
      ) {
        const formattedMinDate = format(
          effectiveMinDate,
          internalFormat ?? "yyyy-MM-dd",
        );
        const formattedMaxDate = format(
          effectiveMaxDate,
          internalFormat ?? "yyyy-MM-dd",
        );
        return `Date should be between ${formattedMinDate.toUpperCase()} - ${formattedMaxDate.toUpperCase()}`;
      }
    } else if (effectiveMinDate) {
      if (validDateStartOfDay < effectiveMinDate) {
        const formattedMinDate = format(
          effectiveMinDate,
          internalFormat ?? "yyyy-MM-dd",
        );
        return `Date must be after ${formattedMinDate.toUpperCase()}`;
      }
    } else if (effectiveMaxDate) {
      if (validDateStartOfDay > effectiveMaxDate) {
        const formattedMaxDate = format(
          effectiveMaxDate,
          internalFormat ?? "yyyy-MM-dd",
        );
        return `Date must be before ${formattedMaxDate.toUpperCase()}`;
      }
    }

    return true;
  };
