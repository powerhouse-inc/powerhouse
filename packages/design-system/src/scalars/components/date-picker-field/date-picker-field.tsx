import {
  DatePicker,
  type DatePickerProps,
} from "../../../ui/components/data-entry/date-picker/date-picker.js";
import { withFieldValidation } from "../fragments/with-field-validation/with-field-validation.js";
import { type FieldErrorHandling } from "../types.js";
import { validateDatePicker } from "./date-picker-validations.js";

interface DatePickerFieldProps extends DatePickerProps, FieldErrorHandling {}

const DatePickerField = withFieldValidation<DatePickerFieldProps>(DatePicker, {
  validations: {
    _timePickerType: validateDatePicker,
  },
});

DatePickerField.displayName = "DatePickerField";

export { DatePickerField, type DatePickerFieldProps };
