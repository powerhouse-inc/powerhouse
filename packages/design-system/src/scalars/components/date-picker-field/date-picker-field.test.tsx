import { screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { renderWithForm } from "../../lib/testing.js";
import { DatePickerField } from "./date-picker-field.js";

vi.mock("#powerhouse", () => ({
  Icon: ({ name, className }: { name: string; className?: string }) => (
    <div data-testid={`mock-icon-${name}`} className={className}>
      Mock Icon: {name}
    </div>
  ),
}));
describe("DatePickerField", () => {
  it("should match the snapshot", () => {
    const { container } = renderWithForm(
      <DatePickerField
        label="Test Label"
        name="test-date"
        value="2025-01-01"
        dateFormat="yyyy-MM-dd"
      />,
    );
    expect(container).toMatchSnapshot();
  });
  it("should display the label when provided", () => {
    const labelText = "Test Label";
    renderWithForm(<DatePickerField name="test-date" label={labelText} />);
    expect(screen.getByText(labelText)).toBeInTheDocument();
  });

  it("should not render the label when label prop is not provided", () => {
    renderWithForm(<DatePickerField name="test-date" />);
    expect(screen.queryByText("Test Label")).not.toBeInTheDocument();
  });

  it("should mark the label as required when required prop is true", () => {
    renderWithForm(
      <DatePickerField name="test-date" label="Test Label" required />,
    );
    const label = screen.getByText("Test Label");
    const asterisk = screen.getByText("*");
    expect(label).toBeInTheDocument();
    expect(asterisk).toBeInTheDocument();
  });

  it("should mark the label as disabled when disabled prop is true", () => {
    renderWithForm(
      <DatePickerField name="test-date" label="Test Label" disabled />,
    );
    const label = screen.getByText("Test Label");
    expect(label).toHaveClass("cursor-not-allowed");
    expect(label).toHaveClass("text-gray-700");
  });

  it("should disable dates before minDate", async () => {
    renderWithForm(
      <DatePickerField
        label="Test Label"
        name="test-date"
        minDate="2025-01-16"
      />,
    );

    // 1. Find and click the calendar button to open it
    const calendarTrigger = screen.getByTestId("mock-icon-CalendarTime");
    await userEvent.click(calendarTrigger);

    // 2. Wait for the calendar dialog to be visible
    const calendar = await screen.findByRole("dialog");

    expect(calendar).toBeInTheDocument();

    // 3. Navigate to previous month (December 2022)
    const prevMonthButton = screen.getByTestId("mock-icon-CaretLeft");
    await userEvent.click(prevMonthButton);

    // 4. Find the date button for February 14, 2025
    const dateButton = screen.getByRole("button", {
      name: "Friday, February 14th, 2025",
    });
    // 5. Check that the date button is disabled
    expect(dateButton).toHaveAttribute("tabIndex", "-1");
    expect(dateButton).toHaveClass("disabled:pointer-events-none");
    // 6. Check that the input is empty
    const input = screen.getByRole("textbox");
    expect(input).toHaveValue("");
  });

  it("should disable dates after maxDate", async () => {
    renderWithForm(
      <DatePickerField
        label="Test Label"
        name="test-date"
        maxDate="2025-01-16"
      />,
    );
    // 1. Find and click the calendar button to open it
    const calendarTrigger = screen.getByTestId("mock-icon-CalendarTime");
    await userEvent.click(calendarTrigger);

    // 2. Wait for the calendar dialog to be visible
    const calendar = await screen.findByRole("dialog");

    expect(calendar).toBeInTheDocument();

    // 3. Navigate to previous month (December 2022)
    const prevMonthButton = screen.getByTestId("mock-icon-CaretLeft");
    await userEvent.click(prevMonthButton);

    const dateButton = screen.getByRole("button", {
      name: "Wednesday, February 26th, 2025",
    });

    // 5. Check that the date button is disabled
    expect(dateButton).toHaveAttribute("tabIndex", "-1");
    expect(dateButton).toHaveClass("disabled:pointer-events-none");
    // 6. Check that the input is empty
    const input = screen.getByRole("textbox");

    expect(input).toHaveValue("");
  });
  // Validate minDate when user add value in the input
  it("should disable dates after minDate when user adds a value in the input", async () => {
    renderWithForm(
      <DatePickerField
        label="Test Label"
        name="test-date"
        minDate="2025-01-16"
        showErrorOnBlur
      />,
    );
    const input = screen.getByRole("textbox");

    expect(input).toBeInTheDocument();
    await userEvent.type(input, "2025-01-10");
    await userEvent.tab();
    expect(screen.getByText(/Date must be on or after/i)).toBeInTheDocument();
  });
});
// Validate maxDate when user add value in the input
it("should disable dates after minDate when user adds a value in the input", async () => {
  renderWithForm(
    <DatePickerField
      label="Test Label"
      name="test-date"
      maxDate="2025-01-16"
      showErrorOnBlur
    />,
  );
  const input = screen.getByRole("textbox");

  expect(input).toBeInTheDocument();
  await userEvent.type(input, "2025-01-20");
  await userEvent.tab();
  expect(screen.getByText(/Date must be on or before/i)).toBeInTheDocument();
});
