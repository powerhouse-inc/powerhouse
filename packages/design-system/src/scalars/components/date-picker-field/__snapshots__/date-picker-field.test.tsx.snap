// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DatePickerField > should match the snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <label
      class="inline-flex items-center text-sm font-medium leading-4 text-gray-900 dark:text-gray-50 mb-[3px]"
      for="test-id"
    >
      Test Label
      <span
        class="ml-1 text-gray-800 group-hover:text-gray-900 dark:text-gray-400 dark:group-hover:text-slate-50"
      >
        *
      </span>
    </label>
    <div>
      <div>
        Placeholder DatePicker
      </div>
    </div>
  </div>
</div>
`;
